src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) 2012-2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "wallet/coinselection.h"
+#include "util.h"
+#include "utilmoneystr.h"
+
+// Descending order comparator
+struct {
+    bool operator()(CInputCoin a, CInputCoin b) const
+    {
+        return a.txout.nValue > b.txout.nValue;
+    }
+} descending;
+
+struct CompareValueOnly
+{
+    bool operator()(const CInputCoin& t1,
+                    const CInputCoin& t2) const
+    {
+        return t1.txout.nValue < t2.txout.nValue;
+    }
+};
+
+bool SelectCoinsBnB(std::vector<CInputCoin>& utxo_pool, const CAmount& target_value, const CAmount& cost_of_change, std::set<CInputCoin>& out_set, CAmount& value_ret)
+{
+    out_set.clear();
+    value_ret = 0;
+
+    if (utxo_pool.size() <=0) {
+        return false;
+    }
+
+    int depth = 0;
+    int tries = 100000;
+    std::vector<std::pair<bool, bool>> selection; // First bool: select the utxo at this index; Second bool: traversing second branch of this utxo
+    selection.assign(utxo_pool.size(), std::pair<bool, bool>(false, false));
+    bool done = false;
+    bool backtrack = false;
+    
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+    
+    // Calculate remaining
+    CAmount remaining = 0;
+    for (CInputCoin utxo : utxo_pool) {
+        remaining += utxo.txout.nValue;
+    }
+    
+    // Depth first search to find 
+    while (!done)
+    {
+        if (value_ret > target_value + cost_of_change) { // Selected value is out of range, go back and try other branch
+            backtrack = true;
+        } else if (value_ret >= target_value) { // Selected value is within range
+            done = true;
+        } else if (tries <= 0) { // Too many tries, exit
```

Comment:
Here's a unexpected behavior in my algorithm: if there is a number of input combinations whose value_ret all exceed the target_value when 'tries == 0' is passed, tries is ignored until the first combination that doesn't exceed target_value.